crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -129,4 +131,146 @@ module sui::random {
+    public fun new_generator(r: &Random, ctx: &mut TxContext): RandomGenerator {
+        let inner = load_inner(r);
+        let seed = hmac_sha3_256(&inner.random_bytes, &to_bytes(fresh_object_address(ctx)));
+        RandomGenerator { seed, counter: 0, buffer: vector[] }
+    }
+
+    // Get the next block of random bytes.
+    fun derive_next_block(g: &mut RandomGenerator): vector<u8> {
+        g.counter = g.counter + 1;
+        hmac_sha3_256(&g.seed, &bcs::to_bytes(&g.counter))
+    }
+
+    // Fill the generator's buffer with 32 random bytes.
+    fun fill_buffer(g: &mut RandomGenerator) {
+        let next_block = derive_next_block(g);
+        vector::append(&mut g.buffer, next_block);
+    }
+
+    /// Generate n random bytes.
+    public fun bytes(g: &mut RandomGenerator, num_of_bytes: u16): vector<u8> {
+        let result = vector[];
+        // Append RAND_OUTPUT_LEN size buffers directly without going through the generator's buffer.
+        let num_of_blocks = num_of_bytes / RAND_OUTPUT_LEN;
+        while (num_of_blocks > 0) {
+            vector::append(&mut result, derive_next_block(g));
+            num_of_blocks = num_of_blocks - 1;
+        };
+        // Take remaining bytes from the generator's buffer.
+        if (vector::length(&g.buffer) < ((num_of_bytes as u64) - vector::length(&result))) {
+            fill_buffer(g);
+        };
+        while (vector::length(&result) < (num_of_bytes as u64)) {
+            vector::push_back(&mut result, vector::pop_back(&mut g.buffer));
+        };
+        result
+    }
+
+    // Helper function that extracts the given number of bytes from the random generator and returns it as u256.
+    // Assumes that the caller has already checked that num_of_bytes is valid.
+    fun u256_from_bytes(g: &mut RandomGenerator, num_of_bytes: u8): u256 {
+        if (vector::length(&g.buffer) < (num_of_bytes as u64)) {
+            fill_buffer(g);
+        };
+        let result: u256 = 0;
+        let i = 0;
+        while (i < num_of_bytes) {
+            let byte = vector::pop_back(&mut g.buffer);
+            result = (result << 8) + (byte as u256);
+            i = i + 1;
+        };
+        result
+    }
+
+    /// Generate a u256.
+    public fun generate_u256(g: &mut RandomGenerator): u256 {
+        u256_from_bytes(g, 32)
+    }
+
+    /// Generate a u128.
+    public fun generate_u128(g: &mut RandomGenerator): u128 {
+        (u256_from_bytes(g, 16) as u128)
+    }
+
+    /// Generate a u64.
+    public fun generate_u64(g: &mut RandomGenerator): u64 {
+        (u256_from_bytes(g, 8) as u64)
+    }
+
+    /// Generate a u32.
+    public fun generate_u32(g: &mut RandomGenerator): u32 {
+        (u256_from_bytes(g, 4) as u32)
+    }
+
+    /// Generate a u16.
+    public fun generate_u16(g: &mut RandomGenerator): u16 {
+        (u256_from_bytes(g, 2) as u16)
+    }
+
+    /// Generate a u8.
+    public fun generate_u8(g: &mut RandomGenerator): u8 {
```

Comment:
Some other APIs that could be useful:

- Generate a random `bool`
- Generate a random `address`
- Generate a random object `ID`
- Generate a random `std::string` (i.e. UTF8 string) of a given length
- Shuffle a `vector<T>` (the whole thing, and/or a selected range)

Reply:
Added `generate_bool` and `shuffle`.
What would be the use cases for a random address, ID or string onchain? (the first two can be easily generated from u256)

Reply:
My thought was just for consistency--if the answer to "how do I generate a random instance of `T`" (where `T` is a native type) is "call the `random::generate_T` function", that seems very intuitive. Whereas if there's sometimes a function and sometimes a need to convert, that is one more step for a dev to think about.